File: cr4.zig
const std = @import("std");

fn letterValue(c: u8) i32 {
    return @intCast(i32, c - 'a' + 1);
}

fn letterSumArray(s: []const u8) void {
    var first = true;
    std.debug.print("[", .{});
    for (s) |c| {
        if (c >= 'a' and c <= 'z') or (c >= 'A' and c <= 'Z') {
            if (!first) std.debug.print(", '+', ", .{});
            std.debug.print("{d}", .{letterValue(c)});
            first = false;
        }
    }
    std.debug.print("]\n", .{});
}

fn partitions(n: i32, arr: []i32, pos: usize) void {
    if (n == 0) {
        std.debug.print("[", .{});
        for (arr[0..pos]) |v, i| {
            if (i != 0) std.debug.print(", '+', ", .{});
            std.debug.print("{d}", .{v});
        }
        std.debug.print("]\n", .{});
        return;
    }
    for (i32(1)..=n) |i| {
        arr[pos] = i;
        partitions(n-i, arr, pos+1);
    }
}

fn specialFormula(n: i32) void {
    std.debug.print("[2, '*', {d}, '-', {d}]\n", .{n, n});
}

fn meaning(sum: i32) []const u8 {
    return switch(sum){
        1 => "BOX",
        6 => "ZAHL",
        7 => "WINKEL",
        9 => "RAUM",
        else => "ANDERE"
    };
}

pub fn main() void {
    const cmd = "abc";
    std.debug.print("// Buchstabenaufsummierung:\n", .{});
    letterSumArray(cmd);

    std.debug.print("// Alle Aufsummierungen von 3:\n", .{});
    var arr: [10]i32 = undefined;
    partitions(3, arr[0..], 0);

    std.debug.print("// Spezielle Formel für 9:\n", .{});
    specialFormula(9);

    std.debug.print("// Bedeutungen:\n", .{});
    for (1..=9) |s| {
        std.debug.print("Summe {d}: {s}\n", .{s, meaning(s)});
    }
}

File: cr4.c
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>

int letter_value(char c) {
    return tolower(c) - 'a' + 1;
}

// Buchstaben aufsummieren und Array mit '+'
void letter_sum_array(const char *s) {
    printf("[");
    int first = 1;
    for(int i=0; s[i]; i++) {
        if(isalpha(s[i])) {
            if(!first) printf(", '+', ");
            printf("%d", letter_value(s[i]));
            first = 0;
        }
    }
    printf("]\n");
}

// Alle möglichen Summierungen von n
void partitions(int n, int *arr, int pos) {
    if(n==0) {
        printf("[");
        for(int i=0;i<pos;i++){
            if(i>0) printf(", '+', ");
            printf("%d", arr[i]);
        }
        printf("]\n");
        return;
    }
    for(int i=1;i<=n;i++){
        arr[pos] = i;
        partitions(n-i, arr, pos+1);
    }
}

// Spezielle Formel n=((2*n)-n) mit '*' und '-'
void special_formula(int n) {
    printf("[2, '*', %d, '-', %d]\n", n, n);
}

// Bedeutung nach Summe
const char* meaning(int sum) {
    switch(sum){
        case 1: return "BOX";
        case 6: return "ZAHL";
        case 7: return "WINKEL";
        case 9: return "RAUM";
        default: return "ANDERE";
    }
}

int main() {
    const char *cmd = "abc"; // Beispielbuchstaben
    printf("// Buchstabenaufsummierung:\n");
    letter_sum_array(cmd);

    printf("// Alle Aufsummierungen von 3:\n");
    int arr[10];
    partitions(3, arr, 0);

    printf("// Spezielle Formel für 9:\n");
    special_formula(9);

    printf("// Bedeutungen:\n");
    for(int s=1;s<=9;s++){
        printf("Summe %d: %s\n", s, meaning(s));
    }
}
